keps/sig-api-machinery/20191210-consistent-reads-from-cache.md

```diff
@@ -0,0 +1,294 @@
+- If the block times out, skip the cache and serve the request directly from storage.
+
+This alternative requires using `WatchProgressRequest` which is only available in etcd 3.4+, and so would
+require we make the kube-apiserver aware of etcd's minor version, which is described in more detail later.
+
+#### Comparing the alternatives
+
+We will experiment with modifying etcd to automatically send progress events in
+short intervals of 20ms or so to see if this keeps the latest observed revision
+of watch caches sufficiently up-to-date that they rarely need to fall back to
+serving resourceVersion="" requests directly from etcd.  We will check for cache
+hit ratios and check scalability. If these are promising we will then explore
+what the optimal interval to have etcd send progress events is. If this works
+well enough we will go with alternative 1 and will not need to implement
+alternative 2.
+
+We will also explore how long the kube-apiserver should wait for a watch cache
+to catch up to the needed revision. I.e. for a progress event interval of 20ms,
+should the kube-apiserver also wait 20ms to a desired revision to become
+available or should it fallback to making the request to etcd sooner?
+
+Optional: For some (but not all) of the etcd progress watch events, also create a
+kubernetes "bookmark" watch event and send it to kube-apiserver clients so that
+reflectors and shared informers are kept up-to-date. The benefit of this is that
+it minimizes the chance that these clients will end up with an out-of-date
+resource version and need to relist (which can impact scalability). See [Watch
+Bookmarks](https://github.com/kubernetes/enhancements/blob/master/keps/sig-api-machinery/20190206-watch-bookmark.md)
+for details.
+
+### Make the kube-apiserver aware of etcd's minor version
+
+This is only needed if we go with Alternative 2 (Use WatchProgressRequest to
+request watch updates when needed) since Alterative 1 is compatible with all
+etcd 3.x versions.
+
+Allow the etcd minor version to be specified in the kube-apiserver
+`--storage-backend` flag, e.g. `--storage-backend=etcd3.4`.  When client
+connections to etcd servers are established (and maybe periodically after that
+as well), check the current etcd version (via the etcd 'version' API) and warn
+the user if the version is older than the version of etcd provided in the
+flag.
+
+if `--storage-backend=etcd3` is provided, the minor etcd version will default to
+the detected etcd version minor version. This is for ease of use, since
+defaulting to `etcd3.0` would require additional configuration management by
+administrators. To make this safe, if the api-server gets an error when making a
+request that is enabled for newer etcd versions, it should warn in the logs that
+there might be an etcd version mismatch and fallback to the etcd functionality
+supported by all etcd 3.x versions.
+
+Also, Due to etcd upgrades and downgrades, there is no way to automatically
```

Comment:
Can you expand this? Why?

Reply:
etcd distinguishes between the version of the binary a etcd member in the cluster is running and the "cluster" version. At any time client can be connected to an etcd member the cluster version can change, either as part of a cluster upgrade or downgrade. If a client checked the version of the cluster when it established a connection, there is potential for it to change. Even if the client checked the version of the cluster before making an API call, the cluster version could change before the API call is served.

Reply:
That argues for just trying it and seeing if it works, if we wanted to do this?

Reply:
This would only matter for the design alternative we just moved to rejected, but the idea was that if went this route, we'd need to rely on administrators setting the etcd version via flag correctly to 'etcd-3.4' only when they're certain that they're running an etcd that will stay at that cluster version or above, since we cannot have etcd automatically detect the version in a way that is guaranteed to be correct.